src/bluesky/callbacks/tiled_writer.py

```diff
@@ -0,0 +1,391 @@
+          - Parse the fields in a new StreamDatum
+          - Increment the number of rows (implemented by the Base class)
+          - Keep track of the correspondence between indices and seq_nums (implemented by the Base class)
+          - Update the list of assets, including their uris, if necessary
+          - Update shape and chunks
+        """
+        self._num_rows += doc["indices"]["stop"] - doc["indices"]["start"]
+        new_seqnums = range(doc["seq_nums"]["start"], doc["seq_nums"]["stop"])
+        new_indices = range(doc["indices"]["start"], doc["indices"]["stop"])
+        self._seqnums_to_indices_map.update(dict(zip(new_seqnums, new_indices)))
+
+    def get_data_source(self) -> DataSource:
+        """Return a DataSource object reflecting the current state of the streamed dataset.
+
+        The returned DataSource is conceptually similar (and can be an instance of) tiled.structures.DataSource. In
+        general, it describes associated Assets (filepaths, mimetype) along with their internal data structure
+        (array shape, chunks, additional parameters) and should contain all information necessary to read the file.
+        """
+        return DataSource(
+            mimetype=self.mimetype,
+            assets=self.assets,
+            structure_family=StructureFamily.array,
+            structure=ArrayStructure(
+                data_type=BuiltinDtype.from_numpy_dtype(self.dtype),
+                shape=self.shape,
+                chunks=self.chunks,
+            ),
+            parameters=self.adapter_parameters,
+            management=Management.external,
+        )
+
+
+class HDF5StreamHandler(StreamHandlerBase):
+    mimetype = "application/x-hdf5"
+
+    def __init__(self, sres: StreamResource, desc: EventDescriptor):
+        super().__init__(sres, desc)
+        self.assets.append(Asset(data_uri=self.uri, is_directory=False, parameter="data_uri"))
+        self.swmr = self._sres_parameters.get("swmr", True)
+
+    @property
+    def adapter_parameters(self) -> Dict:
+        return {"path": self._sres_parameters["path"].strip("/").split("/")}
+
+
+class TIFFStreamHandler(StreamHandlerBase):
+    mimetype = "multipart/related;type=image/tiff"
+
+    def __init__(self, sres: StreamResource, desc: EventDescriptor):
+        super().__init__(sres, desc)
+        self.chunk_size = 1
+        self.data_uris: List[str] = []
+
+    def consume_stream_datum(self, doc: StreamDatum):
+        indx = int(doc["uid"].split("/")[1])
+        new_datum_uri = self.uri.strip("/") + "/" + f"{indx:05d}.tif"
+        new_asset = Asset(
+            data_uri=new_datum_uri,
+            is_directory=False,
+            parameter="data_uris",
+            num=len(self.assets) + 1,
+        )
+        self.assets.append(new_asset)
+        self.data_uris.append(new_datum_uri)
+
+        super().consume_stream_datum(doc)
+
+    @property
+    def adapter_parameters(self) -> Dict:
+        return {"data_uris": self.data_uris}
```

Comment:
ah, you are right! `TiffSequenceAdapter` in Tiled needs to be initialized with `data_uris`, but I think it gets them from the `Assets` directly.
I also agree that `indx:05d.tif` is awful, but I think it needs to be generalized _before_ creating `adapter_parameters` (likely when we declare Assests).

Reply:
I've generalized the filename template for tiff sequences. This relies on the formatting string being passed in the Stream Resource document using Python-specific style (`{:05d}.tif`). This is a bit of limitation of course, but I think it should be okay since the `TIFFStreamHandler` is a custom class and can be redefined for a new use case, if needed.
Please let me know if anyone has a better idea how to handle the sequential filenames in a more general way. Maybe there is a widely accepted standard already (e.g. with %) that we can just convert to Python representation.